istioctl/cmd/istioctl/authn.go

```diff
@@ -0,0 +1,107 @@
+// Copyright 2018 Istio Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"sort"
+	"text/tabwriter"
+
+	"github.com/spf13/cobra"
+)
+
+type AuthenticationDebug struct {
+	Host                     string
+	Port                     int
+	AuthenticationPolicyName string
+	DestinationRuleName      string
+	ServerProtocol           string
+	ClientProtocol           string
+	TLSConflictStatus        string
+}
+
+// can allows user to query Istio RBAC effect for a specific request.
+func check() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "tls_check",
+		Short: "Check whether TLS setting are matching between authentication policy and destination rules",
+		Long: `
+`,
+		Example: ``,
+		Args:    cobra.ExactArgs(0),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			pilots, err := getPilotPods()
+			if err != nil {
+				return err
+			}
+			if len(pilots) == 0 {
+				return errors.New("unable to find any Pilot instances")
+			}
+			if debug, pilotErr := callPilotDiscoveryDebug(pilots, "", "authn"); pilotErr == nil {
+				// fmt.Println(debug)
+				var dat []AuthenticationDebug
+				if err := json.Unmarshal([]byte(debug), &dat); err != nil {
+					panic(err)
+				}
+				sort.Slice(dat, func(i, j int) bool {
```

Comment:
Can we abstract most of this out into `istioctl/pkg` and add some tests. I'm trying to make a conscious effort to have the `cmd/*.go` files as thin as possible going forwards.

Reply:
I think we should refactor pilot debug for this as well (istioctl is just a tool use to query it). I'd like to do it (or someone to do it :) ) in later PRs.